hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotMoveDeletedKeysRequest.java

```diff
@@ -0,0 +1,133 @@
+
+import org.apache.hadoop.ozone.om.OmSnapshot;
+import org.apache.hadoop.ozone.om.OmSnapshotManager;
+import org.apache.hadoop.ozone.om.OzoneManager;
+import org.apache.hadoop.ozone.om.SnapshotChainManager;
+import org.apache.hadoop.ozone.om.helpers.SnapshotInfo;
+import org.apache.hadoop.ozone.om.ratis.utils.OzoneManagerDoubleBufferHelper;
+import org.apache.hadoop.ozone.om.request.OMClientRequest;
+import org.apache.hadoop.ozone.om.request.util.OmResponseUtil;
+import org.apache.hadoop.ozone.om.response.OMClientResponse;
+import org.apache.hadoop.ozone.om.response.snapshot.OMSnapshotMoveDeletedKeysResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.SnapshotMoveKeyInfos;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.SnapshotMoveDeletedKeysRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.apache.hadoop.ozone.om.OmSnapshotManager.getSnapshotPrefix;
+
+/**
+ * Handles OMSnapshotMoveDeletedKeys Request.
+ */
+public class OMSnapshotMoveDeletedKeysRequest extends OMClientRequest {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMSnapshotMoveDeletedKeysRequest.class);
+
+  public OMSnapshotMoveDeletedKeysRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long trxnLogIndex, OzoneManagerDoubleBufferHelper omDoubleBufferHelper) {
+    OmSnapshotManager omSnapshotManager = ozoneManager.getOmSnapshotManager();
+    SnapshotChainManager snapshotChainManager =
+        ozoneManager.getSnapshotChainManager();
+
+    SnapshotMoveDeletedKeysRequest moveDeletedKeysRequest =
+        getOmRequest().getSnapshotMoveDeletedKeysRequest();
+    SnapshotInfo fromSnapshot = SnapshotInfo.getFromProtobuf(
+        moveDeletedKeysRequest.getFromSnapshot());
+
+    // If there is no Non-Deleted Snapshot move the
+    // keys to Active Object Store.
+    SnapshotInfo nextSnapshot = null;
+    OMClientResponse omClientResponse = null;
+    OzoneManagerProtocolProtos.OMResponse.Builder omResponse =
+        OmResponseUtil.getOMResponseBuilder(getOmRequest());
+    try {
+      nextSnapshot = getNextActiveSnapshot(fromSnapshot,
+          snapshotChainManager, omSnapshotManager);
+
+      // Get next non-deleted snapshot.
+
+      List<SnapshotMoveKeyInfos> activeDBKeysList =
+          moveDeletedKeysRequest.getActiveDBKeysList();
+      List<SnapshotMoveKeyInfos> nextDBKeysList =
+          moveDeletedKeysRequest.getNextDBKeysList();
+
+      OmSnapshot omNextSnapshot = null;
+
+      if (nextSnapshot != null) {
+        omNextSnapshot = (OmSnapshot) omSnapshotManager
+            .checkForSnapshot(nextSnapshot.getVolumeName(),
```

Comment:
I think it would better to do this call from within OMSnapshotMoveDeletedKeysResponse.  You could just pass in the SnapshotInfo parameter instead of the OMSnapshot parameter.  What do you think?

Reply:
Yeah it can be done. Thanks for the suggestion, will update the patch.